src/app/clusters/scenes/SceneTable.h

```diff
@@ -0,0 +1,293 @@
+            this->mExtensionFieldSets         = other.mExtensionFieldSets;
+            this->mSceneTransitionTimeSeconds = other.mSceneTransitionTimeSeconds;
+            this->mTransitionTime100ms        = other.mTransitionTime100ms;
+        }
+    };
+
+    /// @brief Struct combining both ID and data of a table entry
+    struct SceneTableEntry
+    {
+
+        // ID
+        SceneStorageId mStorageId;
+
+        // DATA
+        SceneData mStorageData;
+
+        SceneTableEntry() = default;
+        SceneTableEntry(SceneStorageId id) : mStorageId(id) {}
+        SceneTableEntry(const SceneStorageId id, const SceneData data) : mStorageId(id), mStorageData(data) {}
+
+        bool operator==(const SceneTableEntry & other)
+        {
+            return (this->mStorageId == other.mStorageId && this->mStorageData == other.mStorageData);
+        }
+
+        void operator=(const SceneTableEntry & other)
+        {
+            this->mStorageId   = other.mStorageId;
+            this->mStorageData = other.mStorageData;
+        }
+    };
+
+    SceneTable(){};
+
+    virtual ~SceneTable() = default;
+
+    // Not copyable
+    SceneTable(const SceneTable &) = delete;
+
+    SceneTable & operator=(const SceneTable &) = delete;
+
+    virtual CHIP_ERROR Init(PersistentStorageDelegate * storage) = 0;
+    virtual void Finish()                                        = 0;
+
+    // Data
+    virtual CHIP_ERROR SetSceneTableEntry(FabricIndex fabric_index, const SceneTableEntry & entry)                    = 0;
+    virtual CHIP_ERROR GetSceneTableEntry(FabricIndex fabric_index, SceneStorageId scene_id, SceneTableEntry & entry) = 0;
+    virtual CHIP_ERROR RemoveSceneTableEntry(FabricIndex fabric_index, SceneStorageId scene_id)                       = 0;
+    virtual CHIP_ERROR RemoveSceneTableEntryAtPosition(FabricIndex fabric_index, SceneIndex scened_idx)               = 0;
+
+    // SceneHandlers
+    virtual CHIP_ERROR RegisterHandler(SceneHandler * handler)   = 0;
+    virtual CHIP_ERROR UnregisterHandler(SceneHandler * handler) = 0;
+    virtual CHIP_ERROR UnregisterAllHandler()                    = 0;
+
+    // Extension field sets operation
+    virtual CHIP_ERROR SceneSaveEFS(SceneTableEntry & scene)                                    = 0;
+    virtual CHIP_ERROR SceneApplyEFS(FabricIndex fabric_index, const SceneStorageId & scene_id) = 0;
+
+    // Fabrics
+    virtual CHIP_ERROR RemoveFabric(FabricIndex fabric_index) = 0;
+
+    // Iterators
+    using SceneEntryIterator = CommonIterator<SceneTableEntry>;
+
+    virtual SceneEntryIterator * IterateSceneEntry(FabricIndex fabric_index) = 0;
+
+    // Handlers
+    virtual bool HandlerListEmpty() { return (mNumHandlers == 0); }
+    virtual bool HandlerListFull() { return (mNumHandlers >= kMaxSceneHandlers); }
+    virtual uint8_t GetHandlerNum() { return this->mNumHandlers; }
+
+    SceneHandler * mHandlers[kMaxSceneHandlers] = { nullptr };

```

Comment:
Would it make sense to have the handlers be a linked list so we don't have to preallocate memory for them here and have a hard cap and whatnot?

Reply:
> Since the number of handler is limited to the number of "sceneable" clusters,

Where is that limitation coming from?  What I see is that a handler can be registered, it can get called with a given endpoint, then it can return what set of clusters that handler supports for that endpoint, then for each cluster  you ask the handler to save the data it cares about for that cluster for that endpoint.

So you could have an arbitrary number of handlers that all claim to support the same clusters on the same endpoints and save different data, or you could have one handler that handles everything, and everything in between.

I _am_ a little confused by how handlers that overlap each others endpoint/cluster sets are supposed to handle the incoming data blobs, though, now that I look into it again.  So for example how to make sense of the incoming data in `ApplyScene`.

Reply:
Handlers are now stored in linked list, no more maximum handlers per fabric either. Comments explaining how the first handler per <endpoint, cluster> pair added.